manifest/main.js

```diff
@@ -0,0 +1,127 @@
+    `,
+  });
+  argparser.addArgument("--input",  {metavar: "<json>", type: "string", defaultValue: "",
+    help: "Optional input JSON (i.e. the output of a previous run). This is useful for incremental updates"})
+  argparser.addArgument("--local", {action: 'storeTrue',
+    help: 'Access a local copy of S3 inventories within ./devData/. See docstring of fetchInventoryLocal() for expected filenames.'})
+  argparser.addArgument("--restrict-sources", {metavar: "<name>", type: "string", nargs: '+',
+    help: "Only fetch data from a subset of sources. Source names are those defined in DATA_SOURCES"});
+  argparser.addArgument("--output", {metavar: "<json>", required: true})
+  argparser.addArgument("--indent", {action: 'storeTrue', help: 'Indent the output JSON'})
+  argparser.addArgument("--gzip", {action: 'storeTrue', help: 'GZip the output JSON'})
+
+  return argparser.parseArgs();
+}
+
+
+main(parseArgs());
+
+
+async function main(args) {
+
+  // load existing resources from disk, if requested (useful if we update one source at a time)
+  const resources = args.input ?
+    JSON.parse(fs.readFileSync(args.input, {encoding: 'utf8', flag: 'r'})) :
+    {};
+
+  for (const src of DATA_SOURCES) {
+    if (args.restrict_sources && !args.restrict_sources.includes(src.name)) {
+      continue
+    }
+
+    if (src.type !== 's3Inventory') {
+      logger.error(`Data source provided for a type we don't yet handle (${src.type})`)
+    }
+
+    /* Before we fetch the (updated) data and add it into the overall collection of resources,
+    delete any existing data from this data source */
+    for (const resourceType of Object.keys(resources)) {
+      for (const [id, resource] of Object.entries(resources[resourceType])) {
+        if (src.removeResource(resourceType, id, resource)) {
+          delete resources[resourceType][id];
+        }
+      }
+    }
+
+    const groupedObjects = (await parseInventory(
+      {bucket: src.bucket, prefix: src.prefix, name: src.name, local: args.local}
+    ))
+      .map(src.categorise)
+      .filter((item) => !!item)
+      // Collect together all items ("files") based on their assigned resourceType & id
+      .reduce((store, item) => {
+        const {resourceType, id} = item;
+        if (!store[resourceType]) store[resourceType] = {};
+        if (!store[resourceType][id]) store[resourceType][id] = [];
+        store[resourceType][id].push(item);
+        return store;
+      }, {})
+
+    // Turn those groups of files (each with the same ID) into a resource object, including
+    // versions etc and add the object into the main `resources` data structure.
+    for (const resourceType of Object.keys(groupedObjects)) {
+      const resourcesById = Object.fromEntries(
+        Object.entries(groupedObjects[resourceType])
+          .map(([id, items]) => [id, src.createResource(resourceType, id, items)])
+          .filter((resource) => !!resource)
```

Comment:
The `map` here always produces values like `[id, someObject]`, so how would those end up ever being falsey and getting removed by the `filter`...? What am I missing?

Reply:
That should have been `.filter(([, resource]) => !!resource)`. I've modified the code to be a bit more explicit. (Conceptually, `createResource` can create a resource with no valid versions, which we want to filter.)